### ISSUE TITLE
Problems when using custom learning rate scheduler for NeuralForecast Models

### ISSUE THREAD

**Issue by MLfreakPy** [Nixtla/neuralforecast#1096]
### What happened + What you expected to happen

Hi together, 

first, thank you so much for the implementation of a custom lr scheduler!

I ran into problems employing a `ReduceLROnPlateau ` scheduler.

Currently I am passing through the `lr_scheduler` and  `lr_scheduler_kwargs` parameters. However, I get the below error when implementing ReduceLROnPlateau. Other schedulers, like OneCycle, work perfectly though!  I think some of your earlier discussion evolved around it, however, I couldn't make it work yet.
[(https://github.com/Nixtla/neuralforecast/pull/998)]

`MisconfigurationException('The lr scheduler dict must include a monitor when a ReduceLROnPlateau scheduler is used. For example: {"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "monitor": "your_loss"}}').`

### Versions / Dependencies

GoogleColab
neuralforecast
Version: 1.7.4

### Reproduction script

See above - if needed happy to supply more details.

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by MLfreakPy** [comment]
Maybe @JQGoh, @jmorales, @BrunoBelucci you have some idea of how to solve it?

**Comment by marcopeix** [comment]
Hello! Can you send a code snippet of OneCycle that works and one for ReduceLROnPlateau to reproduce on my end? Thanks!

**Comment by BrunoBelucci** [comment]
Hello,

To the best of my knowledge, it is not currently possible to use ReduceLROnPlateau out-of-the-box with the existing implementation of neuralforecast. As mentioned in [this comment](https://github.com/Nixtla/neuralforecast/pull/998#issuecomment-2126773941), this would require the user to provide a "lr_scheduler_config," which is not directly supported at this time. This "lr_scheduler_config" should be returned by the configure_optimizers method in Lightning (refer to this [documentation](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#configure-optimizers) for more details).

In the "lr_scheduler_config," you need to specify the "monitor" parameter with the metric you're tracking to detect the plateau. Typically, the "interval" is also set to "epoch" because the plateau is usually identified per epoch rather than per step, though this may vary depending on your model.

Currently, I am not using neuralforecast with ReduceLROnPlateau, so I cannot offer an immediate solution. However, based on our previous discussion in #998, I believe the simplest workaround for now is to monkey-patch the configure_optimizers function of your model to return a dictionary as expected by Lightning. This dictionary should include an "optimizer" key and, optionally, a "lr_scheduler" key, which can be a single LR scheduler or a lr_scheduler_config (as specified in the documentation).

If you do not manage to solve this, you can provide me a minimum (not) working example and I can see if I can fix it for you during my free time.

Regards,
Bruno

**Comment by MLfreakPy** [comment]
Thank you so much for your explanations and looking into how to resolve the issue!!

Below a code snippet that (a) works with OneCycleLR and (b) that reproduces the error with ReduceLRonPlateau.
```
# GENERAL + DATA-PROCESSING
!pip install neuralforecast
from neuralforecast import NeuralForecast
from neuralforecast.models import NHITS
from neuralforecast.utils import AirPassengersDF

Y_df = AirPassengersDF # Defined in neuralforecast.utils
Y_df.head()
```

a.  - OneCycleLR
```
from torch.optim.lr_scheduler import OneCycleLR
      
lr_scheduler_cls = OneCycleLR
max_lr = 1e-2
lr_scheduler_kwargs = {
    'max_lr': max_lr,
    'total_steps': 50}

horizon = 12

models = [ NHITS(h=horizon,                   
                input_size=2 * horizon,     
                max_steps=50,               
                lr_scheduler = lr_scheduler_cls, 
                lr_scheduler_kwargs = lr_scheduler_kwargs) ]
nf = NeuralForecast(models=models, freq='M')
nf.fit(df=Y_df)
```

b. -ReduceLROnPlateau

```

from torch.optim.lr_scheduler import ReduceLROnPlateau
lr_scheduler_cls = ReduceLROnPlateau
lr_scheduler_kwargs = {
      'mode': 'min',  
      'factor': 0.5,
      'patience': 2}

horizon = 12

models = [ NHITS(h=horizon,                   
                input_size=2 * horizon,     
                max_steps=50,               
                lr_scheduler = lr_scheduler_cls, 
                lr_scheduler_kwargs = lr_scheduler_kwargs) ]
nf = NeuralForecast(models=models, freq='M')
nf.fit(df=Y_df)
```
ERROR MESSAGE

> [/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/optimizer.py](https://localhost:8080/#) in _configure_schedulers_automatic_opt(schedulers, monitor)
>     275             )
>     276             if scheduler["reduce_on_plateau"] and scheduler.get("monitor", None) is None:
> --> 277                 raise MisconfigurationException(
>     278                     "The lr scheduler dict must include a monitor when a `ReduceLROnPlateau` scheduler is used."
>     279                     ' For example: {"optimizer": optimizer, "lr_scheduler":'
> 
> MisconfigurationException: The lr scheduler dict must include a monitor when a `ReduceLROnPlateau` scheduler is used. For example: {"optimizer": optimizer, "lr_scheduler": {"scheduler": scheduler, "monitor": "your_loss"}}

**Comment by MLfreakPy** [comment]
Thank you very much @JQGoh!! I will use this dev-version and your example for implementation 👍 This flexibility in configuring the optimizers is amazing :)

**Comment by MLfreakPy** [comment]
@JQGoh I tried to manipulate the `set_configure_optimizers ` as you described. For that reason I (tried to) install the _developer version_ of nixtla as per below. Maybe I do this wrongly. I cannot see the `set_configure_optimizers ` parameter in the imported neuralforecast models nor use it to implement LR on plateau. Do you have an idea where the problem resides?

```
# install DEV-version of nixtla (yet not sucessfull?)
!git clone https://github.com/Nixtla/neuralforecast.git
%cd neuralforecast
!pip install -e

#
from torch.optim.lr_scheduler import ReduceLROnPlateau
optimizer = torch.optim.Adam(LSTMm.parameters(), lr=learning_rate)
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer = optimizer, mode='min', factor=0.5, patience=2)

model.set_configure_optimizers(optimizer=optimizer,scheduler=scheduler,monitor="train_loss")

```

**Comment by jmoralez** [comment]
Hey. I think we can introduce that method and deprecate the arguments

**Comment by MLfreakPy** [comment]
@jmoralez That sounds great😊
